pkg/controller/pod.go

```diff
@@ -202,15 +214,92 @@ func (c *Controller) syncPod(ctx context.Context, pod *corev1.Pod) error {
 		}
 	}
 
-	klog.V(4).Infof("Registering instance %s under service %+v", pod.Name, cloudmapConfig)
+	cloudMapServiceKey := *cloudmapConfig.NamespaceName + "-" + *cloudmapConfig.ServiceName
+
+	existingItem, exists, _ := c.cloudMapInstanceCache.Get(&cloudMapInstanceCacheItem{
+		key: cloudMapServiceKey,
+	})
+
+	var serviceInstanceSummary map[string]bool
+
+	if exists {
+		if healthy, ok := existingItem.(*cloudMapInstanceCacheItem).instanceSummary[pod.Status.PodIP]; ok {
+			//Right now we're not logging pod/instance health info to CloudMap. So, if it is in cache then it is
+			//considered healthy. *TODO - v1.0* */
+			klog.V(4).Info("Pod Instance already registered. Nothing to do for this pod; Healthy?: ", healthy)
+			return nil
+		}
+		serviceInstanceSummary = existingItem.(*cloudMapInstanceCacheItem).instanceSummary
+		if serviceInstanceSummary == nil { serviceInstanceSummary = make(map[string]bool) }
+		serviceInstanceSummary[pod.Status.PodIP] = true
+	} else {
+		//Retrieve CloudMap instances for this service
+		serviceInstanceSummary := make(map[string]bool)
+		registered := false
+
+		appmeshCloudMapConfig := &appmesh.AwsCloudMapServiceDiscovery{
+			NamespaceName: cloudmapConfig.NamespaceName,
+			ServiceName:   cloudmapConfig.ServiceName,
+		}
+
+		if serviceInstances, _ := c.getServiceInstancesFromCloudMap(ctx, appmeshCloudMapConfig); len(serviceInstances) > 0 {
```

Comment:
`getServiceInstancesFromCloudMap` does not utilize a cache. Does that mean we're calling `ListInstances` fully paginated for every single pod?

Reply:
Key for the cacheItem is service name...So, we get all the instances per that service from cloud map on a cache miss and populate the cache. Tried it out with 50 pods per service, and no additional registerinstance calls were made on a cache miss (assuming there is no change in instance id)

Reply:
Ah, I misread what this was doing, my bad. Got confused since it's intermingled with the registration behavior. Would be nice if this was structured more like a read-through cache that populates itself on a miss rather than the caller being responsible for it. If something like this will exist in V2, then that's a good place for it.

Also, you're referring to `RegisterInstance` but I'm specifically referring to `ListInstances` here, as we don't want to be wasteful on any of the APIs.